Draft overview for custom_expectations (#3911)
great-expectations/great_expectations: docs/guides/expectations/creating_custom_expectations/overview.md

---
title: Overview
---

> import Prerequisites from './components/prerequisites.jsx'

Reviewer: It seems this could be a more broadly shared component rather than being local to create_custom_expectations

Author: Agreed. I actually copied it from `docs/guides/connecting_to_your_data/components/prerequisites.jsx`.

There are a few different versions of this already. AFACIT, they differ only in the defaults that they include. These defaults tend to differ by area within the ToC, so it makes sense to have them split out by subdirectory in the docs.

I'd suggest doing a separate PR to refactor all these `Prerequisites` components to use a common base class.

You can extend the functionality of Great Expectations by creating your own custom Expectations. You can also enrich Great Expectations as a shared standard for data quality by contributing new Expectations to the open source project.

> Both processes follow the same streamlined steps. This section will teach you how.

Reviewer 2: I'm not certain this is accurate.  It seems to me that creating a custom expectation is one process, and contributing a created expectation to the open source is another process that is dependent on the creation process having been completed. Possibly you mean that creating a custom expectation and creating and contributing a custom expectation follow the same steps -- but that's still a bit inaccurate since the second process would have additional steps that the first stopped before reaching.

I would change this to: These processes compliment each other and their steps are streamlined so that one flows into the other.  Once you have created a custom expectation it is simple to contribute it to the open source project, and this section will teach you how to do both.

Author: ```suggestion
These processes compliment each other and their steps are streamlined so that one flows into the other. Once you have created a custom Expectation, it is simple to contribute it to the open source project. This section will teach you how to do both.
```

> These processes compliment each other and their steps are streamlined so that one flows into the other. Once you have created a custom Expectation, it is simple to contribute it to the open source project. This section will teach you how to do both.

Reviewer 3: I think it should be "complement"

Author: ```suggestion
These processes complement each other and their steps are streamlined so that one flows into the other. Once you have created a custom Expectation, it is simple to contribute it to the open source project. This section will teach you how to do both.
```

In order to make development of Expectations as easy as possible, we've broken up the steps to create custom Expectations into a series of bite-sized steps. Each step can be completed in minutes. They can be completed (and contributed) incrementally, unlocking value at each step along the way.

Grouped together, they constitute a Definition of Done for Expectations at each [Level of Maturity](/docs/contributing/contributing_maturity).

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.0.0-beta3/css/all.min.css" crossorigin="anonymous" referrerpolicy="no-referrer"/>
<i class="fas fa-circle" style={{color: "#dc3545"}}></i> An experimental Expectation...

* Has a `library_metadata` object
* Has a docstring, including a one-line short description
* Has at least one positive and negative example case
* Has core logic and passes tests on at least one `Execution Engine`

<i class="fas fa-circle" style={{color: "#ffc107"}}></i> A beta Expectation...

* Has basic input validation and type checking
* Has all four statement `Renderers`: question, descriptive, prescriptive, diagnostic
* Has default `Parameter Builders` and Domain hooks to support Profiling   

> * Has core logic exists and passes tests for all applicable `Execution Engines` and SQL dialects

Reviewer 2: "Has core logic exists and passes" reads as though a couple of variations of the statement were composed and part of the losing variation wasn't removed.  I'd remove the word 'exists.'

Author: ha. You're totally right.
```suggestion
* Has core logic that passes tests for all applicable `Execution Engines` and SQL dialects
```

<i class="fas fa-check-circle" style={{color: "#28a745"}}></i> A production Expectation...

> * Passes all linting checks

Reviewer 3: Does it make sense to have a bullet for "Has thorough input validation and type checking, as determined by project code standards" ? It's a little vague, but I like the idea of encouraging more of this prior to elevating to production.

Author: We have three other things that are similar in spirit:

* Has basic input validation and type checking (at the beta level)
* Passes all linting checks (production)
* Has passed a manual review by a code owner for code standards and style guides (production)

I figured these would cover the same substantive material. If you want to emphasize input validation and type checking, I suppose we could add a bullet, but it seems redundant to me.

* Has all applicable Renderers, with fully typed and styled output
* Has a full suite of tests, as determined by project code standards
* Has passed a manual review by a code owner for code standards and style guides

## How these docs are organized

> The docs in `Creating custom Expectations` focus on completing the four steps for experimental Expectations. Completing them will get to the point where your Expectation can be executed against one backend, with a couple tests to verify correctness, and a basic docstring and metadata to support diagonstics. Optionally, you can also publish experimental Expectations to the [Great Expectations open source gallery](https://greatexpectations.io/expectations) by following the steps [here](overview#publishing-your-expectation-as-an-open-source-contribution).

Reviewer 3: - "couple **of** tests"
- "diagonstics" is misspelled.

Author: ```suggestion
The docs in `Creating custom Expectations` focus on completing the four steps for experimental Expectations. Completing them will get to the point where your Expectation can be executed against one backend, with a couple of tests to verify correctness, and a basic docstring and metadata to support diagnostics. Optionally, you can also publish experimental Expectations to the [Great Expectations open source gallery](https://greatexpectations.io/expectations) by following the steps [here](overview#publishing-your-expectation-as-an-open-source-contribution).
```

> These code to achieve the first four steps looks somewhat different depending on the class of Expectation you're developing. Accordingly, there are separate how-to guides and templates for each class of Expectation.

Reviewer 2: "These code" -> "The code"

> These code to achieve the first four steps looks somewhat different depending on the class of Expectation you're developing. Accordingly, there are separate how-to guides and templates for each class of Expectation.

Author: ```suggestion
The code to achieve the first four steps looks somewhat different depending on the class of Expectation you're developing. Accordingly, there are separate how-to guides and templates for each class of Expectation.
```

| Has all four statement `Renderers`: question, descriptive, prescriptive, diagnostic              | [How to add text renderers for custom Expectations](../features_custom_expectations/how_to_add_text_renderers_for_an_expectation) |
| Has default `Parameter Builders` and Domain hooks to support Profiling                           | |
| Has core logic exists and passes tests for all applicable `Execution Engines` and SQL dialects   | [How to add SQLAlchemy support for custom Metrics](../features_custom_expectations/how_to_add_sqlalchemy_support_for_an_expectation)<br/> [How to add Spark support for custom Metrics](../features_custom_expectations/how_to_add_spark_support_for_an_expectation)|
| Passes all linting checks                                                                        | |
| Has all applicable Renderers, with fully typed and styled output                                 | |
| Has a full suite of tests, as determined by project code standards                               | |
| Has passed a manual review by a code owner for code standards and style guides                   | |

## Publishing your Expectation as an open source contribution

> (Add this later; possibly in a different location in the TOC)

Reviewer: Could you link to: contributing/contributing for now?

Author: Sure, although these docs won't go live until after another merge from `hackathon-docs` to `develop`.

Author: ```suggestion
You can find more detailed instructions in the [Contributing](/docs/contributing/contributing) section of the docs.
```
